### PR TITLE
Patch param modules earlier

### PR DESCRIPTION
**Pull request by ezr-ondrej**
We can end up with controllers getting required prior to_prepare and
that breaks patching the host_params.

Refs #132

### CODE REVIEW

**Comment by ezr-ondrej** [comment]
@laugmanuel could you check with your setup it fixes #132?

**Comment by laugmanuel** [comment]
@ezr-ondrej it does not seem to help. Now I do see the Puppetclasses in GUI (still with the same layout problem), but I can't see them using `rake console`.
Removing a Puppetclass using GUI is also still not possible.

`rake console`:
```ruby
[46] pry(main)> Host.find_by(name: 'gail-galluzzo.foo.example.local').puppetclasses
=> []
````

API:
```sh
❯ curl -u admin:admin http://localhost:5000/api/hosts/gail-galluzzo.foo.example.local | jq '.puppetclasses'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  3451    0  3451    0     0   7042      0 --:--:-- --:--:-- --:--:--  7028
[
  {
    "id": 2,
    "name": "test",
    "module_name": "test"
  },
  {
    "id": 1,
    "name": "test::config",
    "module_name": "test"
  }
]

❯ curl -u admin:admin http://localhost:5000/api/hosts/gail-galluzzo.foo.example.local/puppetclasses | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   151    0   151    0     0    480      0 --:--:-- --:--:-- --:--:--   480
{
  "total": 0,
  "subtotal": 0,
  "page": 1,
  "per_page": 20,
  "search": null,
  "sort": {
    "by": null,
    "order": null
  },
  "results": {}
}
```

**Comment by ezr-ondrej** [comment]
Yeah didn't realize, we are breaking the current methods, that will be dropped and API redirected to new enpoints, but your finds should work for now.

**Comment by ezr-ondrej** [comment]
What about now, I didn't want pollute the Host model by puppet related methods much, but I'm fine with it for 2.5 :)
And it should fix your finds.

**Review by laugmanuel — APPROVED** [review]
This works like a charm now :) Together with #143 this seems to resolve all issues found in #132

File: app/models/concerns/foreman_puppet/extensions/host.rb [hunk 1]
```diff
@@ -61,6 +61,12 @@ def search_by_config_group(_key, operator, value)
       end
 
       module PrependedMethods
+        # TODO: we can drop this once extracted_from_core?
+        def validate_association_taxonomy(association_name)
+          return if association_name.to_sym == :environment
+          super
+        end
```
Comment (kamils-iRonin) [inline app/models/concerns/foreman_puppet/extensions/host.rb:68]:
```suggestion
        # TODO: we can drop this once extracted_from_core?
        unless ForemanPuppet.extracted_from_core?
          def validate_association_taxonomy(association_name)
            return if association_name.to_sym == :environment
            super
          end
        end
```

Comment (ezr-ondrej) [reply app/models/concerns/foreman_puppet/extensions/host.rb:68]: we can do that, but I'd rather drop this method once we extract puppet from core.
If we accidently don't do that, this should be `noop`, so I feel the check is not necessary.

Comment (kamils-iRonin) [reply app/models/concerns/foreman_puppet/extensions/host.rb:68]: if we add the `puppet_extracted` core branch to CI, won't there be a problem with tests without this check?

Comment (ezr-ondrej) [reply app/models/concerns/foreman_puppet/extensions/host.rb:68]: Nope, this method will be noop

Comment (ezr-ondrej) [reply app/models/concerns/foreman_puppet/extensions/host.rb:68]: there would be no such association on the `Host::Managed`

**Pull request merged**